Providing --path option to transfer-ownership

sharidas opened the pull request:
This will help user to selectively move the folders
specified using --path option, instead of moving
entire folder under files directory.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
The change has been made for occ command's transfer-ownership. An additional --path argument has been added to support. This will help user to pass particular folder/file to move to destination user/target. So I have introduced an optional argument --path. I have deliberately made it optional, as its user choice to move entire folder under files or selectively.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
This issue which this patch addresses is https://github.com/owncloud/platform/issues/91

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This helps the user to selectively transfer the ownership of the files/folders instead of moving entire folder under files.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Below is the way I have tested:

The folder files/1/1a/1aa/ has 1aaa, 1aab and 1aac folders
root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud# sudo -u www-data ./occ  files:transfer-ownership --path "test1/files/1/1a/1aa" test1 test2 
Analysing files of test1 ...
    0 [>---------------------------]
Collecting all share information for files and folder of test1 ...
    0 [>---------------------------]
Transferring files to test2/files/transferred from test1 on 20170309_101206 ...
Restoring shares ...
    0 [>---------------------------]
root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud#

after the execution I was able to see in test2 user:

root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud# ls -lt data/test2/files/transferred\ from\ test1\ on\ 20170309_101206/
total 12
drwxr-xr-x 2 www-data www-data 4096 Mar  9 15:42 1aac
drwxr-xr-x 2 www-data www-data 4096 Mar  9 15:42 1aab
drwxr-xr-x 2 www-data www-data 4096 Mar  9 15:42 1aaa
root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud#

and in test1 user:

root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud# ls data/test1/files/1/1a/
1ab
root@silburyadmin-Inspiron-5567:/home/sujith/test/owncloud#

<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Providing --path option to transfer-ownership

This will help user to selectively move the folders
specified using --path option, instead of moving
entire folder under files directory.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -86,6 +89,12 @@ protected function configure() {
+				'path',
+				'',
+				InputArgument::OPTIONAL,
+				'selectively provide the path to copy. For example --path "alice/files/Music"'
```
s/copy/transfer/ (it's a move, not a copy)

sharidas replied:
Replaced copy with transfer in the new update.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -214,9 +226,12 @@ private function collectUsersShares(OutputInterface $output) {
-		$view->mkdir("$this->sourceUser/files");
+		$sourcePath = (strlen($this->inputPath) > 0) ? $this->inputPath : "$this->sourceUser/files";
+		$view->rename($sourcePath, $this->finalTarget);
+		if (!is_dir($this->sourceUser/files)) {
```
missing double quotes on string ?

sharidas replied:
Fixed the double quotes on string in the new update.

sharidas commented:
The updated version of the patch is ready for review. @PVince81  Can you please review this change.

PVince81 commented:
- [ ] BUG: specifying a non-existing path on the CLI must fail instead of attempting transfer

- [ ] BUG: some problem with transferring shares, steps:

1. Create three users "user1", "user2" and "user3"
1. From "user3" share a folder "foruser1" with "user1"
1. Login as "user1"
1. Create a folder "transferthis"
1. Move "foruser1" to "transferthis"
1. Create a folder "transferthis/foruser2/x"
1. Share folder "transferthis/foruser2" with "user2" AND "user3"
1. Create a folder "transferthis/normal/y"
1. Log out
1. ```
± % occ files:transfer-ownership user1 user2 --path=transferthis
Analysing files of user1 ...
    0 [>---------------------------]
Collecting all share information for files and folder of user1 ...
    2 [============================]
Transferring files to user2/files/transferred from user1 on 20170310_021750 ...
Restoring shares ...
 1/2 [==============>-------------]  50%Share with id 3 points at deleted file, skipping
 2/2 [============================] 100%
```
1. Login as "user1" and "user2" to check.

Expected result: folder "transferthis" is no more in "user1"'s account and is present in "user2's" account.
Actual result: nothing was transferred.

Please have a look.

PVince81 commented:
you might need to debug into `$view->rename` to find out what happens and whether it fails somewhere there

sharidas commented:
@PVince81  Thanks for the feedback. I will have a look at it.

PVince81 commented:
Best would be to write an integration test right away because setting this up for testing is time consuming as you need to start from scratch after a transfer.

Running the test when debugging would help save this time.

PVince81 commented:
@sharidas it seems you have found a bug deep in the FS code.

In the old code, the move operation would do: `rename('test1/files', 'test2/files/transferred ...')` which moves "files" to the target and renames it to "transferred ...". So far so good, this is legitimate.

But now with the new logic the rename is: `rename('test1/files/a', 'test2/files/transferred .../a')`.
However the folder "transferred ..." doesn't exist yet.

Still, it seems the `View` happily moves the folder to this non-existing target folder instead of failing... resulting in the inconsistency you saw. The filecache will contain "a" but with a parent -1 because the parent did not exist / have an id.

@sharidas for you the fix is to call `$view->mkdir()` for the target folder first.

The View but will need to be fixed separately, I'll make a ticket for it.

PVince81 commented:
Raised View issue here: https://github.com/owncloud/core/issues/27376

sharidas commented:
@PVince81 I have updated the PR by adding $view->mkdir at https://github.com/owncloud/core/pull/27343/files#diff-0674cc8bda871e32d63705af16fa8144R258. I have verified that share folders and other folders which are created by user1 do get transferred.

sharidas committed:
Update the integration test for transfer-ownership

Update the integration test for transfer-ownership
as the new option --path is introduced in the command.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

SergioBertolinSG commented on tests/integration/features/transfer-ownership.feature:
```diff
@@ -114,6 +114,16 @@ Feature: transfer-ownership
+		Given user "user0" exists
+		And user "user1" exists
+		And User "user0" created a folder "/test"
+		When transfering ownership path from path "test" "user0" to "user1"
```
This sentence is not clear. 
Something like this seems better:

When transfering ownership of path "test" from "user0" to "user1"

Also I recommend to put the starting slash before any paths as it is the decided choice to name paths in scenarios.

sharidas commented:
@SergioBertolinSG I have updated the integration tests. Kindly review the same.

SergioBertolinSG commented on tests/integration/features/transfer-ownership.feature:
```diff
@@ -126,3 +126,126 @@ Feature: transfer-ownership
+		And user "user1" exists
+		And User "user0" created a folder "/test"
+		When transfering ownership of path "test" from "user0" to "user1"
+		And the command was successful
```
Missing Then

SergioBertolinSG commented:
Tests passed locally. 


But this explanation in the help of the command is wrong:

Options:
      --path=PATH       selectively provide the path to transfer. For example --path "alice/files/Music"

Which this changes It has to be:
--path PATH

Or change it to use '=' with the command.

sharidas commented:
@PVince81 @SergioBertolinSG I have made the changes. Kindly review them.

PVince81 commented:
Seems to have trouble with PHP 5.6 ?

```
± % sudo -u wwwrun ./occ
PHP Fatal error:  Class 'OCA\Files\Command\InputOption' not found in /srv/www/htdocs/owncloud/apps/files/lib/Command/TransferOwnership.php on line 96
PHP Stack trace:
PHP   1. {main}() /srv/www/htdocs/owncloud/occ:0
PHP   2. require_once() /srv/www/htdocs/owncloud/occ:11
PHP   3. OC\Console\Application->loadCommands() /srv/www/htdocs/owncloud/console.php:98
PHP   4. OC\Console\Application->loadCommandsFromInfoXml() /srv/www/htdocs/owncloud/lib/private/Console/Application.php:108
PHP   5. OC\ServerContainer->query() /srv/www/htdocs/owncloud/lib/private/Console/Application.php:166
PHP   6. OC\AppFramework\Utility\SimpleContainer->query() /srv/www/htdocs/owncloud/lib/private/ServerContainer.php:80
PHP   7. OC\AppFramework\Utility\SimpleContainer->resolve() /srv/www/htdocs/owncloud/lib/private/AppFramework/Utility/SimpleContainer.php:113
PHP   8. OC\AppFramework\Utility\SimpleContainer->buildClass() /srv/www/htdocs/owncloud/lib/private/AppFramework/Utility/SimpleContainer.php:92
PHP   9. ReflectionClass->newInstanceArgs() /srv/www/htdocs/owncloud/lib/private/AppFramework/Utility/SimpleContainer.php:75
PHP  10. OCA\Files\Command\TransferOwnership->__construct() /srv/www/htdocs/owncloud/lib/private/AppFramework/Utility/SimpleContainer.php:75
PHP  11. Symfony\Component\Console\Command\Command->__construct() /srv/www/htdocs/owncloud/apps/files/lib/Command/TransferOwnership.php:76
PHP  12. OCA\Files\Command\TransferOwnership->configure() /srv/www/htdocs/owncloud/lib/composer/symfony/console/Command/Command.php:63
```

I did run 'make clean` and `make` to make sure it wasn't a deps issue.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -86,6 +89,12 @@ protected function configure() {
+				'path',
+				null,
+				InputOption::VALUE_OPTIONAL,
+				'selectively provide the path to transfer. For example --path "alice/files/Music"'
```
the example is wrong, it should not contain the "$userId/files/" piece

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +112,26 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+			$objectsDir = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($path));
+			$dirFound = false;
+			foreach ($objectsDir as $name => $object) {
+				if (strpos($name, $this->inputPath)) {
```
is this safe ? what if there are two directories within the same path, like "somedir/test/somedir" ?

Can you add a PHP comment to explain why you are using `strpos` here ?

Would also be good to add an explicit condition like `>= 0`

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +112,26 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		if (strlen($this->inputPath) >= 1) {
+			$unknownDir = $this->inputPath;
+			$this->inputPath = $this->sourceUser . "/files/" . $this->inputPath;
+			$path = realpath(getcwd());
```
Argh, no. Don't do anything with the direct FS. There are some storages that do not store anything in the "data" folder but store in object store.

Please always use the `$view` methods for checking path existence. `$view->is_dir($this->inputPath)`.

In general in OC never ever access the FS directly, always use a $view or the files public API.

sharidas commented:
@PVince81 Kindly review my updated PR. I have removed ugly changes which directly access FS. Used $view->is_dir in the updated PR. I have also modified the example 'selectively provide the path to transfer. For example --path "folder_name" or --path="folder_name'. 
The reason for failure in 5.6 was not having "use Symfony\Component\Console\Input\InputOption;". I hav e fixed that too.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -86,6 +90,12 @@ protected function configure() {
+				'path',
+				null,
+				InputOption::VALUE_OPTIONAL,
+				'selectively provide the path to transfer. For example --path "folder_name" or --path="folder_name"'
```
`--path "folder_name" or --path="folder_name"` ? it's twice the same

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +113,18 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+			$unknownDir = $this->inputPath;
+			$this->inputPath = $this->sourceUser . "/files/" . $this->inputPath;
+			if (!$view->is_dir($this->inputPath)) {
+				$output->writeln("<error>Unknown path provided:$unknownDir</error>");
```
add a space after ":" for better readability

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +113,18 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		if (strlen($this->inputPath) >= 1) {
+			$view = new View();
+			$unknownDir = $this->inputPath;
+			$this->inputPath = $this->sourceUser . "/files/" . $this->inputPath;
```
I suggest you use `ltrim($this->inputPath, '/')` to avoid doubling slashes in case the user input a leading slash in the path

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +113,18 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		$this->inputPath = $input->getOption('path');
+
+		if (strlen($this->inputPath) >= 1) {
+			$view = new View();
```
Before you can use the view you need to first call `initMountPoints` which is done a bit later.
I suggest you move this validation after the `initMountPoints` call below, or at least the one from the source user.

The reason for this is that sometimes the storage is not directly on FS but in an object store. Only if the object store is mounted for that user will you be able to access it with the View.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -214,9 +243,20 @@ private function collectUsersShares(OutputInterface $output) {
+		// This change will help user to transfer the folder specified using --path option.
+		// Else only the content inside folder is transferred which is not correct.
+		if (strlen($this->inputPath) > 0) {
+			if($this->inputPath !== "$this->sourceUser/files") {
```
ah yes, when comparing paths you really need to make sure there are no double slashes.

this makes the `ltrim($path, '/')` call I suggested above really necessary.

PVince81 reviewed (commented):
Almost there!

A few more required changes, especially the one about init mount points.

sharidas commented:
@PVince81 I have moved the validation part for the path under the initMount. Have provided ltrim for the path ( i.e, $this->inputPath ) and during the path comparison I have used ltrim both in the left hand side and right hand side of the comparison. I have added leading space for $unknownDir ( around the </error>). I have removed --path="folder_name" in the help. All these are made in the updated commit.

PVince81 commented:
Integration tests fail, see https://jenkins.owncloud.org/job/owncloud-core/job/core/job/PR-27343/15/display/redirect

SergioBertolinSG commented:
This behaviour should not be this way:

```
/occ files:transfer-ownership --help

--path[=PATH] selectively provide the path to transfer. For example --path "folder_name"
```

it looks like path value is optional, that's misleading.

It should look like

```
./occ files:transfer-ownership       

                                                                    
  [Symfony\Component\Console\Exception\RuntimeException]            
  Not enough arguments (missing: "source-user, destination-user").  
                                                                    

files:transfer-ownership [--path [PATH]] [--] <source-user> <destination-user>
```
Which is what appears when running without required parameters.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -103,6 +113,8 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		$this->sourceUser = $sourceUserObject->getUID();
 		$this->destinationUser = $destinationUserObject->getUID();
+		$this->inputPath = $input->getOption('path');
+		$this->inputPath = ltrim($this->inputPath);
```
`ltrim($this->inputPath,'/')` to remove trailing slashes

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -214,9 +243,20 @@ private function collectUsersShares(OutputInterface $output) {
+		// This change will help user to transfer the folder specified using --path option.
+		// Else only the content inside folder is transferred which is not correct.
+		if (strlen($this->inputPath) > 0) {
+			if(ltrim($this->inputPath, '/') !== ltrim("$this->sourceUser/files", '/')) {
```
you don't need to ltrim the input path if you do it right at the start

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -118,6 +130,16 @@ protected function execute(InputInterface $input, OutputInterface $output) {
+		if (strlen($this->inputPath) >= 1) {
+			$view = new View();
+			$unknownDir = $this->inputPath;
+			$this->inputPath = $this->sourceUser . "/files/" . $this->inputPath;
```
the reason I'm asking you to ltrim the path is that if someone sets `--path=/somedir`, this concatenation here would result in "$userId/files//somedir" (double slash)

PVince81 commented:
:+1: if tests still pass
